### ISSUE TITLE
Querystrings in stylesheet resource are removed when written to tag

### ISSUE THREAD

**Issue by aalcutt** [OrchardCMS/Orchard#7314]
Inside of ResourceDefinition.ResolveUrl() I added logic to append a querystring to the url to act as versioning. By the time the execution gets to ResourceManager.WriteResource() the url of a stylesheet will not contain the querystring while script files will.

```
...
 var path = HostingEnvironment.MapPath(url);
 var file = new FileInfo(path);
 var fileVersion = file.Exists ? file.LastWriteTime.ToString("yyyyMMddhhmm") : "none";
 var version = $"v={HttpUtility.UrlEncode(fileVersion)}";
 version = (path.Contains("?") ? "&" : "?") + version;
 url +=  version;
...
```

Setting the url when defining the resource results will result in the same issue.

```
manifest.DefineScript("main").SetUrl("main.min.js?v=1", "main.js?v=1")
 manifest.DefineStyle("main").SetUrl("main.min.css?v=1", "main.css?v=1")
```

**Comment by rtpHarry** [comment]
I have encountered issues with this as well with a site recently. I thought assigning a [version to the resource](https://github.com/OrchardCMS/Orchard/blob/6720b71cf3474a9a7b8a8cc9a99d58b1e733acfa/src/Orchard/UI/Resources/ResourceDefinition.cs#L83) added a query string like your second example. 

In the site I was testing it with it didn't do that though but I was sure I'd seen it somewhere so maybe the site had some other module that was overriding that behaviour.

**Comment by aalcutt** [comment]
Yeah I noticed the version property doesn't do anything for versioning the url it is just used by the resource manager for breaking ties between resources with the same name.

**Comment by rtpHarry** [comment]
I know in the past there were some issues with using querystrings. I think some browsers refused to cache them at all. I don't know if this is still a valid concern.

I guess my real question is what is the official cache-breaking strategy for Orchard when you push out a new version of the site?

**Comment by sebastienros** [comment]
Maybe add an explicit `SetQuery` to the resource, just in changing the default behavior could be breaking. Or an option `PreserveQueryString()`.

@Piedone do you know why the query string is stripped by default?

**Comment by Piedone** [comment]
I don't know and I can't think of a valid reason; if what @rtpHarry said was the original cause then that's long not an issue.

I'd just silently change the behavior. If you reference a resource with a query string-containing URL then it should continue to work.

**Comment by aalcutt** [comment]
I ended up moving some of the logic into the WriteResource method so it is appended at the last possible moment.

I think the query string is getting lost in StylesheetBindingStrategy on line 97 when it uses the relative url. Why do stylesheets having a custom binding strategy and scripts don't? I don't fully understand what the class is doing.

**Comment by Piedone** [comment]
The point of that is to be able to override stylesheets like shape templates. E.g. if you have a site.css in your base them then you can just put a site.css in the derived them and the latter file will be loaded.

I can't recall the reason why this is like as is but somebody, maybe Sebastien explained it once. I also opened at least one issue about it but I could only find this one which lacks any explanation: https://github.com/OrchardCMS/Orchard/issues/2656

**Comment by aalcutt** [comment]
Interesting, but does that mean scripts cannot override parent theme scripts?
